### PR TITLE
feat(store) added dynamodb store

### LINKED ISSUE #8: DynamoDB support
Any plans to support DynamoDB ? Would you be open for a contribution with adding support for DynamoDB ?

### PR DESCRIPTION
**Pull request by wolfeidau**
closes #8

### CODE REVIEW

**Comment by victorcoder** [comment]
Thanks @wolfeidau could you split the PR in one for bbolt and this one for Dynamo please? so we can focus on just Dynamo here.

**Comment by wolfeidau** [comment]
Yeah got a bit carried away with fixing issues, and enhancing the dynamodb code.

@victorcoder 👍  breaking it up now into another now, will post a PR then rebase it out of this change.

Thanks!

**Comment by wolfeidau** [comment]
I have updated this in line with #12 so it should shrink if this is merged.

**Comment by wolfeidau** [comment]
Happy to work on locking, was planning to do that next, just wanted to get feedback on the initial changes. 

I have worked on and with the Java implementation before there is also this library https://github.com/intercom/lease to learn from.

**Comment by nnarendravijay** [comment]
@wolfeidau @victorcoder There is also a go implementation of locking that I used and have tested with. The leadership election process works pretty well with it when a node dies. https://github.com/hashicorp/vault/tree/master/physical/dynamodb

**Comment by wolfeidau** [comment]
@nnarendravijay That is a great link, very interesting with the metrics, logging and code structure.

A lot to learn from it, thanks for posting!

**Comment by abronan** [comment]
@wolfeidau That's an amazing contribution! Thanks! I'm going to spend some time reviewing this as I've been taking a bit of a pause regarding Open Source work.

There is just a small issue with `gofmt` that must be super easy to solve in order to pass the build.

File: store/dynamodb/dynamodb.go [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+	}
+
+	return ddb.retryDeleteTree(items)
+}
+
+// AtomicPut Atomic CAS operation on a single value.
+func (ddb *DynamoDB) AtomicPut(key string, value []byte, previous *store.KVPair, options *store.WriteOptions) (bool, *store.KVPair, error) {
+
+	getRes, err := ddb.getKey(key, &store.ReadOptions{
+		Consistent: true, // enable the read consitant flag
```
Comment (abronan) [inline store/dynamodb/dynamodb.go]: very minor typo in comment: consitant -> consistent :)

File: store/dynamodb/dynamodb.go [hunk 2]
```diff
@@ -0,0 +1,719 @@
...
+	}
+
+	return true, item, nil
+}
+
+// AtomicDelete delete of a single value
+func (ddb *DynamoDB) AtomicDelete(key string, previous *store.KVPair) (bool, error) {
+
+	getRes, err := ddb.getKey(key, &store.ReadOptions{
+		Consistent: true, // enable the read consitant flag
```
Comment (abronan) [inline store/dynamodb/dynamodb.go]: Same small typo here

File: store/dynamodb/dynamodb.go [hunk 3]
```diff
@@ -0,0 +1,719 @@
...
+
+	success, err := l.tryLock(lockHeld, stopChan)
+	if err != nil {
+		return nil, err
+	}
+	if success {
+		return lockHeld, nil
+	}
+
+	// this really needs a jitter for backoff
```
Comment (abronan) [inline store/dynamodb/dynamodb.go]: Append the comment with a `FIXME:` or `TODO:` so we could track this easily.

File: store/dynamodb/dynamodb_test.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+	assert.NoError(t, err)
+	assert.Equal(t, &store.KVPair{Key: "abc123", Value: []uint8{0x61, 0x62, 0x63, 0x31, 0x32, 0x33, 0xa}, LastIndex: 0xa}, kv)
+
+	data[encodedValueAttribute] = &dynamodb.AttributeValue{S: aws.String("not base64")}
+	kv, err = decodeItem(data)
+	assert.Error(t, err)
+	assert.Nil(t, kv)
+}
+
+func TestPutWithTTL(t *testing.T) {
```
Comment (abronan) [inline store/dynamodb/dynamodb_test.go]: The `TestPutWithTTL` is already in `utils` so we could remove it in DynamoDB tests?

**Comment by wolfeidau** [comment]
@abronan As requested I have corrected the issues you highlighted and now the build is green :tada:

File: store/dynamodb/dynamodb.go [hunk 5]
```diff
@@ -0,0 +1,719 @@
...
+	// ErrLockAcquireCancelled stop called before lock was acquired.
+	ErrLockAcquireCancelled = errors.New("stop called before lock was acquired")
+)
+
+// Register register a store provider in valkeyrie for AWS DynamoDB
+func Register() {
+	valkeyrie.AddStore(store.DYNAMODB, New)
+}
+
+// New opens a creates a new table
```
Comment (victorcoder) [inline store/dynamodb/dynamodb.go]:
```suggestion
// New opens and creates a new table
```

File: store/dynamodb/dynamodb.go [hunk 6]
```diff
@@ -0,0 +1,719 @@
...
+	DeleteTreeTimeoutSeconds = 30
+
+	partitionKey          = "id"
+	revisionAttribute     = "version"
+	encodedValueAttribute = "encoded_value"
+	ttlAttribute          = "expiration_time"
+	lockAttribute         = "lock"
+
+	noExpiration   = time.Duration(0)
+	defaultLockTTL = 60 * time.Second
```
Comment (victorcoder) [inline store/dynamodb/dynamodb.go]: `defaultLockTTL` is 20s in consul and etcd stores should we use the same here?

File: store/dynamodb/dynamodb.go (hunk 5 above)
Comment (victorcoder) [reply store/dynamodb/dynamodb.go]: @wolfeidau After testing I noticed that `createTable` is not used and the table should be created by hand. It is not that I don't like it, but what did you miss a call or it is intended and the comment is not correct?

**Comment by victorcoder** [comment]
I tested the code with a real application and it works pretty well, congrats! I'm really interested in merge this PR, @wolfeidau just take a look at my minor comments and I think we are ready to go.

Thanks!

**Comment by wolfeidau** [comment]
@victorcoder Great to hear it is useful, I have made the changes you suggested.

File: store/dynamodb/dynamodb.go (hunk 5 above)
Comment (victorcoder) [reply store/dynamodb/dynamodb.go]: Thanks for fixing the typo @wolfeidau, but what about my question, should the table be created in the call to `New`?

**Comment by abronan** [comment]
Sorry for the delay again, but I think we're good to go :) LGTM

Last thing @wolfeidau before proceeding: Could you please squash your commits and sign it? I added automated checks on PR to ensure these steps are respected before merging. Thanks!

**Comment by victorcoder** [comment]
Additionally @wolfeidau, could you rebase with master and include the new dependencies in `go.mod`? Thanks

**Comment by wolfeidau** [comment]
I have squashed commits and signed it 👍

**Comment by abronan** [comment]
🎉

**Pull request merged**